### ISSUE TITLE
Elaborate a strategy to launch boosters in OSIO

### ISSUE THREAD

**Issue by gastaldi** [fabric8-launcher/launcher-planning#119]
Right now OSIO supports only the simple boosters. We need to define a strategy on how to make the other boosters to run in OSIO. 
The other Boosters require installing services first, so that may need to be something to be added to the final Jenkinsfiles

**Comment by gastaldi** [comment]
The strategy we came up is to grab the Jenkinsfile that exists in the booster repository, extract the stages we find beween `scm checkout` and `stage ("Deploy")` and insert it in the new Jenkinsfile below the existing `scm checkout` line

**Comment by edewit** [comment]
A nicer way would be to have a special Jenkinsfile that gets included by the pipeline, but that would require us to change all the boosters and all the pipelines

**Comment by quintesse** [comment]
> beween scm checkout and stage ("Deploy")

between `scm checkout` and `stage ("Build")` actually

**Comment by gastaldi** [comment]
@tako then this file doesn't apply: https://github.com/snowdrop/spring-boot-configmap-booster/blob/master/Jenkinsfile

**Comment by gastaldi** [comment]
Probably we'd need to check `Build` first and fallback to `Deploy` as a second option

**Comment by quintesse** [comment]
Or we need to change them and use the same name everywhere. After all it's only a label so people could have chosen whatever text they liked.

**Comment by gastaldi** [comment]
Right, or we could just tell them to put a fragment of that file in the `.openshiftio/` folder and make our lives easier :)

**Comment by edewit** [comment]
So tried it out on prod with the vert.x configmap booster:

Adding the role gave the following error and stopped the build:
```
[tx-configmap-booster_master-72FN25H45BJYZPN3PDHNODBRQVUFGFGJJWX6SHQY62NRVVCMW7GQ] Running shell script
+ oc policy add-role-to-user view -z default
Error from server: the server does not allow access to the requested resource (get policyBindings :default)
```
So tried again without adding the role, hoping that it might not be needed got the following:
```
[tx-configmap-booster_master-72FN25H45BJYZPN3PDHNODBRQVUFGFGJJWX6SHQY62NRVVCMW7GQ] Running shell script
+ oc get configmap app-config -o yaml
+ grep app-config.yml
Error from server: configmaps "app-config" not found
+ oc create configmap app-config --from-file=app-config.yml
configmap "app-config" created
```
So that seemed to work, but it created the configmap in the `edewit-osio-jenkins` project instead of in the `edewit-osio-run` project or `edewit-osio-staging`
We might need a better plan

**Comment by Ladicek** [comment]
AFAIK, the `Jenkinsfile` isn't a mandatory part of the booster and e.g. the WildFly Swarm boosters no longer include it.

**Comment by edewit** [comment]
@Ladicek sure, it's definitely not the greatest solution, but it also doesn't work it started as an idea

**Comment by Ladicek** [comment]
FWIW, including a snippet of a `Jenkinsfile` (or just a shell script?) in `.openshiftio` sounds like a reasonable idea to me.

**Comment by edewit** [comment]
Thinking we needed to add something to the specific environment I tried one more time [creating the configmap here](https://github.com/edewit/vertx-configmap-booster/blob/master/Jenkinsfile#L31)

**Comment by edewit** [comment]
right @Ladicek the problem is only to find a place in the pipeline Jenkinsfile to call this script or add this snippet

**Comment by Ladicek** [comment]
@edewit this line https://github.com/edewit/vertx-configmap-booster/blob/master/Jenkinsfile#L31 you linked above runs in the `mavenNode`. I'd say you need to inject something like

```groovy
clientsNode {
  sh ".............."
}
```

right in between `mavenNode` and `if (utils.isCD()){`. The `clientNode` should have `oc` available.

**Comment by edewit** [comment]
In order to install something like a config map into the "stage" namespace we need to update the oc command version otherwise a command like this will ignore the `-n` flag
```
  sh "oc create -n ${envStage} configmap app-config --from-file=app-config.yml" 
```
current version on prod is 1.4, with a newer command this works:
```
sh "curl -L https://github.com/openshift/origin/releases/download/v3.7.1/openshift-origin-client-tools-v3.7.1-ab0f056-linux-64bit.tar.gz | tar -xz"  
  sh "if ! ./openshift-origin-client-tools-v3.7.1-ab0f056-linux-64bit/oc get -n ${envStage} configmap app-config -o yaml | grep app-config.yml; then ./openshift-origin-client-tools-v3.7.1-ab0f056-linux-64bit/oc create -n ${envStage} configmap app-config --from-file=app-config.yml; fi" 
```

**Comment by edewit** [comment]
only problem that remains after this is that the app is not authorized to read the configMap

**Comment by Ladicek** [comment]
There's an `oc` command for that as well: `oc policy add-role-to-user view -z default`. It's not needed for WildFly Swarm, fyi.

**Comment by miteshvp** [comment]
@edewit @gastaldi - any updates here? Any working PR/Branch that I can tag along?

**Comment by edewit** [comment]
I think we're close but since Friday all jenkins builds are stuck on prod
so we have to wait on that being fixed

**Comment by krishnapaparaju** [comment]
@edewit Is this been moved to production ? This is critical fix unblocking couple of boosters to work at OSIO.

**Comment by gastaldi** [comment]
@krishnapaparaju we haven't fixed this issue yet, we'll close this issue once it is resolved in prod-preview, so stay tuned 😉

**Comment by miteshvp** [comment]
@gastaldi @edewit - I see PRs for wfswarm and springboot has been merged. Can it be tested on prod?

**Comment by gastaldi** [comment]
No, because the required PR inhttps://github.com/fabric8io/fabric8-jenkinsfile-library/pull/31 wasn't merged yet
